### PR TITLE
Test dir of pandas.api.*

### LINKED ISSUE #52688: TST: Test dir of pandas.api.*
We currently do not test the `dir` of submodules of `pandas.api`. In #48578, a test is being added for `pandas.api.typing`. Tests like this should be added for each of the submodules - extensions, indexers, interchange, and types.

### PR DESCRIPTION
**Pull request by SecretLake**
- [x] Closes #52688
- [x] Tests added and passed
- [x] All code checks passed
- [x] Added type annotations

@rhshadrach Could you please review this PR?

### CODE REVIEW

File: pandas/tests/api/test_api.py [hunk 1]
```diff
@@ -17,10 +23,12 @@ def check(self, namespace, expected, ignored=None):
         result = sorted(
             f for f in dir(namespace) if not f.startswith("__") and f != "annotations"
         )
+
```
Comment (rhshadrach) [inline pandas/tests/api/test_api.py:26]: Can you remove this added line. I think the code is fine either way (someone could argue to remove the blank lines with the justification that they unnecessarily use up vertical space), and in such cases it's best to leave as-is.

File: pandas/tests/api/test_api.py [hunk 2]
```diff
@@ -17,10 +23,12 @@ def check(self, namespace, expected, ignored=None):
...
         if ignored is not None:
             result = sorted(set(result) - set(ignored))
 
         expected = sorted(expected)
+
```
Comment (rhshadrach) [inline pandas/tests/api/test_api.py:31]: Same

File: pandas/tests/api/test_api.py [hunk 3]
```diff
@@ -237,8 +245,14 @@ def test_depr(self):
 
 
 class TestApi(Base):
-    allowed = ["types", "extensions", "indexers", "interchange", "typing"]
-    allowed_typing = [
+    allowed_api_dirs: list[str] = [
```
Comment (rhshadrach) [inline pandas/tests/api/test_api.py:248]: By-and-large, we do not type-hint test code. I think we should keep the status-quo here unless an explicit decision is made to add type-hints. Can you remove throughout.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! Looks good - some minor requests.

File: pandas/tests/api/test_api.py (hunk 1 above)
Comment (SecretLake) [reply pandas/tests/api/test_api.py]: Absolutely, thank you for the review. Implemented all the changes.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @SecretLake!

**Pull request merged**